### PR TITLE
eclass/tree-sitter-grammar: fix ABI autodetection

### PR DESCRIPTION
**Pull request by msva**
CC: @00-matt @sarnex ?
// PR as result of a conversation on IRC to make all interested people to discuss it here

### CODE REVIEW

File: eclass/tree-sitter-grammar.eclass [hunk 1]
```diff
@@ -42,7 +42,9 @@ EXPORT_FUNCTIONS src_compile src_install
 # This internal function determines the ABI version of a grammar library based
 # on the package version.
 _get_tsg_abi_ver() {
-	if ver_test -gt 0.21; then
+	if [[ -n ${TS_ABI} ]]; then
```
Comment (thesamesam) [inline eclass/tree-sitter-grammar.eclass:45]: Please [document](https://devmanual.gentoo.org/eclass-writing/index.html#documenting-eclasses) this variable using eclassdoc tags.

**Comment by thesamesam** [comment]
Thinking about it, please explain in the commit message why this is important & why it would be needed / when you should use it.

**Comment by 00-matt** [comment]
Currently the tree-sitter ABI stuff is a bit half-baked as all of the grammars currently in ::gentoo are compatible with the version of libtree-sitter.

Do you know which grammar doesn't line up with the current versioning scheme?

I was originally planning on reading the version as a symbol in the library, but it looks like we can coerce it out of the `parser.c(c)` file. Should have time to do that today, unless you wanted to have a go at doing it? :)

**Comment by msva** [comment]
@00-matt @thesamesam 
take a look at new ABI-version detection way (that we discussed on IRC), please

File: eclass/tree-sitter-grammar.eclass [hunk 2]
```diff
@@ -42,13 +42,7 @@ EXPORT_FUNCTIONS src_compile src_install
 # This internal function determines the ABI version of a grammar library based
 # on the package version.
```
Comment (00-matt) [inline eclass/tree-sitter-grammar.eclass:43]:
```suggestion
# on a constant in the source file.
```

File: eclass/tree-sitter-grammar.eclass [hunk 3]
```diff
@@ -42,13 +42,7 @@ EXPORT_FUNCTIONS src_compile src_install
...
 _get_tsg_abi_ver() {
-	if ver_test -gt 0.21; then
-		die "Grammar too new; unknown ABI version"
-	elif ver_test -ge 0.19.0; then
-		echo 13
-	else
-		die "Grammar too old; unknown ABI version"
-	fi
+	sed -r -n '/#define LANGUAGE_VERSION ([0-9]*)$/{s@@\1@;p}' "${S}"/parser.c
```
Comment (00-matt) [inline eclass/tree-sitter-grammar.eclass:45]: Hmm this didn't work for me, also I think we need to die on failure.

I tried to simplify it a bit and ended up with this:

```suggestion
	sed -n '/#define LANGUAGE_VERSION/s/.* //p' "${S}"/parser.c ||
	  die 'Failed to extract ABI version'
```

**Review by 00-matt — CHANGES_REQUESTED** [review]
Thank you for working on this.

If it works for all of the tree-sitter-* packages I think it's worth sending it off to gentoo-dev ml for final review :)

File: eclass/tree-sitter-grammar.eclass (hunk 3 above)
Comment (msva) [reply eclass/tree-sitter-grammar.eclass:45]: > this didn't work for me

Uhm... Strange... I tested it on multiple grammars (like `-c`, `-haskell` before commiting. But ok, I'll apply your suggested changes in few minutes.

Comment (sarnex) [reply eclass/tree-sitter-grammar.eclass:45]: +1 on the die

Comment (sarnex) [inline eclass/tree-sitter-grammar.eclass:45]: Can we at least add a comment explaining the regex? It's obviously much less readable than ver_test

Comment (msva) [reply eclass/tree-sitter-grammar.eclass:45]: @00-matt said that regex failed for him and suggested another way (which is also much more readable), I'll test and apply it in meantime.

**Comment by msva** [comment]
I've applied suggestions from @00-matt.

Is it anything else?

**Comment by thesamesam** [comment]
If sarnex and matt are happy, please send to the gentoo-dev ML with:
```git send-email --to="gentoo-dev@lists.gentoo.org" --cc="sarnex@gentoo.org" --cc="matt@offtopica.uk" HEAD~1```

I'd appreciate it if you could add a bit more information to the commit message about why it's needed / what purpose it serves (what was wrong with the old approach?). Things like this can help folks in future.

If you've not used ```git send-email``` before, you might need this in ~/.gitconfig:
```
[sendemail]
       	smtpEncryption = tls
       	smtpServer = smtp.gentoo.org
       	smtpUser = mva@gentoo.org
       	smtpServerPort = 587
       	annotate = yes
```

File: eclass/tree-sitter-grammar.eclass [hunk 4]
```diff
@@ -40,15 +40,10 @@ EXPORT_FUNCTIONS src_compile src_install
...
+# on a constant in the source file.
 _get_tsg_abi_ver() {
-	if ver_test -gt 0.21; then
-		die "Grammar too new; unknown ABI version"
-	elif ver_test -ge 0.19.0; then
-		echo 13
-	else
-		die "Grammar too old; unknown ABI version"
-	fi
+	sed -n '/#define LANGUAGE_VERSION/s/.* //p' "${S}"/parser.c ||
```
Comment (sarnex) [inline eclass/tree-sitter-grammar.eclass:45]: I don't understand what this sed command is doing, I could look it up, but that kind of defeats the point. Can we add a comment?

**Review by sarnex — APPROVED** [review]
With that one thing fixed, lgtm

**Comment by msva** [comment]
@sarnex I've added description comment
@thesamesam I've added verbose decription in commit message

please, approve

**Review by sarnex — APPROVED** [review]

**Review by thesamesam — APPROVED** [review]
Happy if sarnex & matt are, next step is posting to ML.

**Pull request closed without merging**